lib/services.cf

```diff
@@ -218,6 +219,16 @@ bundle agent standard_services(service,state)
       "$(init) $(state)"
       contain => silent;
 
+    sysvservice.freebsd.start.!enabled::
+      "/usr/sbin/sysrc $(service)_enable=YES"
+      handle => "standard_services_sysvservice_enable",
+      comment => "If the service should be running, make sure it is enabled";
```

Comment:
You are absolutely right. It should restricted to sysvservice instead of chkconfig to make sense. The last commit should do the trick - can you verify that?

Reply:
I will check and update when done.

Reply:
Review the state of the env at start of manual test

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # pkg info nginx
    pkg: No package(s) matching nginx
    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # service nginx onestatus
    nginx does not exist in /etc/rc.d or the local startup
    directories (/usr/local/etc/rc.d), or is not executable
    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cat services/main.cf 
    ###############################################################################
    #
    # bundle agent main
    #  - User/Site policy entry
    #
    ###############################################################################
    
    bundle agent main
    # User Defined Service Catalogue
    {
      methods:
        # Activate your custom policies here
    
      packages:
        "nginx"
          policy => "absent",
          package_module => pkg;
    
      services:
        "nginx" service_policy => "stop";
    
    }

Modifying the policy to promise nginx package is present.

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # sed -i .back 's/absent/present/' services/main.cf        
    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cat services/main.cf
    ###############################################################################
    #
    # bundle agent main
    #  - User/Site policy entry
    #
    ###############################################################################
    
    bundle agent main
    # User Defined Service Catalogue
    {
      methods:
        # Activate your custom policies here
    
      packages:
        "nginx"
          policy => "present",
          package_module => pkg;
    
      services:
        "nginx" service_policy => "stop";
    
    }
    
Run policy to see if nginx package state changes as expected.

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cf-agent -KIf update.cf; cf-agent -KI
        info: Copied file '/usr/local/share/examples/cfengine-masterfiles/masterfiles/services/main.cf' to '/var/cfengine/inputs/services/main.cf.cfnew' (mode '600')
        info: Removed old backup '/var/cfengine/inputs/services/main.cf.cfsaved'
        info: Backed up '/var/cfengine/inputs/services/main.cf' as '/var/cfengine/inputs/services/main.cf.cfsaved'
        info: Moved '/var/cfengine/inputs/services/main.cf.cfnew' to '/var/cfengine/inputs/services/main.cf'
        info: Updated '/var/cfengine/inputs/services/main.cf' from source '/usr/local/share/examples/cfengine-masterfiles/masterfiles/services/main.cf' on 'localhost'
        info: files promise '/var/cfengine/inputs' repaired
        info: Installing cfe_internal_non_existing_package...
        info: Q:pkg install -y CF_NO ...:Updating FreeBSD repository catalogue...
        info: Q:pkg install -y CF_NO ...:FreeBSD repository is up to date.
        info: Q:pkg install -y CF_NO ...:All repositories are up to date.
        info: Q:pkg install -y CF_NO ...:pkg: No packages available to install matching 'CF_NOMATCH-CF_NOMATCH' have been found in the repositories
        info: Successfully installed package 'nginx'
        info: Executing 'no timeout' ... '/usr/sbin/sysrc nginx_enable=NO'
      notice: Q: "...sbin/sysrc ngin": nginx_enable: YES -> NO
        info: Last 1 quoted lines were generated by promiser '/usr/sbin/sysrc nginx_enable=NO'
        info: Completed execution of '/usr/sbin/sysrc nginx_enable=NO'
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation

Checking on status of nginx package after policy run, expected package to be present now.

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # pkg info nginx
    nginx-1.18.0_47,2
    Name           : nginx
    Version        : 1.18.0_47,2
    Installed on   : Thu Mar  4 16:25:38 2021 UTC
    Origin         : www/nginx
    Architecture   : FreeBSD:12:amd64
    Prefix         : /usr/local
    Categories     : www
    Licenses       : BSD2CLAUSE
    Maintainer     : joneum@FreeBSD.org
    WWW            : https://nginx.org/
    Comment        : Robust and small WWW server
    Options        :
            AJP            : off
            ARRAYVAR       : off
            AWS_AUTH       : off
            BROTLI         : off
            CACHE_PURGE    : off
            CLOJURE        : off
            CT             : off
            DEBUG          : off
            DEBUGLOG       : off
            DEVEL_KIT      : off
            DRIZZLE        : off
            DSO            : on
            DYNAMIC_HC     : off
            DYNAMIC_TLS    : off
            DYNAMIC_UPSTREAM: off
            ECHO           : off
            ENCRYPTSESSION : off
            FILE_AIO       : on
            FORMINPUT      : off
            GOOGLE_PERFTOOLS: off
            GRIDFS         : off
            GSSAPI_HEIMDAL : off
            GSSAPI_MIT     : off
            HEADERS_MORE   : off
            HTTP           : on
            HTTPV2         : on
            HTTPV2_AUTOTUNE: off
            HTTP_ACCEPT_LANGUAGE: off
            HTTP_ADDITION  : on
            HTTP_AUTH_DIGEST: off
            HTTP_AUTH_JWT  : off
            HTTP_AUTH_KRB5 : off
            HTTP_AUTH_LDAP : off
            HTTP_AUTH_PAM  : off
            HTTP_AUTH_REQ  : on
            HTTP_CACHE     : on
            HTTP_DAV       : on
            HTTP_DAV_EXT   : off
            HTTP_EVAL      : off
            HTTP_FANCYINDEX: off
            HTTP_FLV       : on
            HTTP_FOOTER    : off
            HTTP_GEOIP2    : off
            HTTP_GUNZIP_FILTER: on
            HTTP_GZIP_STATIC: on
            HTTP_IMAGE_FILTER: off
            HTTP_IP2LOCATION: off
            HTTP_IP2PROXY  : off
            HTTP_JSON_STATUS: off
            HTTP_MOGILEFS  : off
            HTTP_MP4       : on
            HTTP_MP4_H264  : off
            HTTP_NOTICE    : off
            HTTP_PERL      : off
            HTTP_PUSH      : off
            HTTP_PUSH_STREAM: off
            HTTP_RANDOM_INDEX: on
            HTTP_REALIP    : on
            HTTP_REDIS     : off
            HTTP_RESPONSE  : off
            HTTP_REWRITE   : on
            HTTP_SECURE_LINK: on
            HTTP_SLICE     : on
            HTTP_SLICE_AHEAD: off
            HTTP_SSL       : on
            HTTP_STATUS    : on
            HTTP_SUB       : on
            HTTP_SUBS_FILTER: off
            HTTP_TARANTOOL : off
            HTTP_UPLOAD    : off
            HTTP_UPLOAD_PROGRESS: off
            HTTP_UPSTREAM_CHECK: off
            HTTP_UPSTREAM_FAIR: off
            HTTP_UPSTREAM_STICKY: off
            HTTP_VIDEO_THUMBEXTRACTOR: off
            HTTP_XSLT      : off
            HTTP_ZIP       : off
            ICONV          : off
            IPV6           : on
            LET            : off
            LINK           : off
            LUA            : off
            MAIL           : on
            MAIL_IMAP      : off
            MAIL_POP3      : off
            MAIL_SMTP      : off
            MAIL_SSL       : on
            MEMC           : off
            MODSECURITY3   : off
            NAXSI          : off
            NJS            : off
            OPENTRACING    : off
            PASSENGER      : off
            POSTGRES       : off
            RDS_CSV        : off
            RDS_JSON       : off
            REDIS2         : off
            RTMP           : off
            SET_MISC       : off
            SFLOW          : off
            SHIBBOLETH     : off
            SLOWFS_CACHE   : off
            SMALL_LIGHT    : off
            SRCACHE        : off
            STREAM         : on
            STREAM_SSL     : on
            STREAM_SSL_PREREAD: on
            THREADS        : on
            VOD            : off
            VTS            : off
            WEBSOCKIFY     : off
            WWW            : on
            XSS            : off
    Shared Libs required:
            libpcre.so.1
    Annotations    :
            FreeBSD_version: 1202000
            cpe            : cpe:2.3:a:nginx:nginx:1.18.0:::::freebsd12:x64:47
            repo_type      : binary
            repository     : FreeBSD
    Flat size      : 1.35MiB
    Description    :
    NGINX is a high performance edge web server with the lowest memory footprint
    and the key features to build modern and efficient web infrastructure.
    
    NGINX functionality includes HTTP server, HTTP and mail reverse proxy, caching,
    load balancing, compression, request throttling, connection multiplexing and
    reuse, SSL offload and HTTP media streaming.
    
    WWW: https://nginx.org/
    WWW: https://nginx.com/
    
Checking on the status of the nginx service, expected to be not running

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # service nginx onestatus
    nginx is not running.
    
    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cat services/main.cf
    ###############################################################################
    #
    # bundle agent main
    #  - User/Site policy entry
    #
    ###############################################################################
    
    bundle agent main
    # User Defined Service Catalogue
    {
      methods:
        # Activate your custom policies here
    
      packages:
        "nginx"
          policy => "present",
          package_module => pkg;
    
      services:
        "nginx" service_policy => "stop";
    
    }
    
Modifying test policy to promise the service should be running...

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # sed -i .bak 's/stop/start/' services/main.cf
    
Running the agent to see if the service gets started as desired...

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cf-agent -KIf update.cf ; cf-agent -KI
        info: Copied file '/usr/local/share/examples/cfengine-masterfiles/masterfiles/services/main.cf' to '/var/cfengine/inputs/services/main.cf.cfnew' (mode '600')
        info: Removed old backup '/var/cfengine/inputs/services/main.cf.cfsaved'
        info: Backed up '/var/cfengine/inputs/services/main.cf' as '/var/cfengine/inputs/services/main.cf.cfsaved'
        info: Moved '/var/cfengine/inputs/services/main.cf.cfnew' to '/var/cfengine/inputs/services/main.cf'
        info: Updated '/var/cfengine/inputs/services/main.cf' from source '/usr/local/share/examples/cfengine-masterfiles/masterfiles/services/main.cf' on 'localhost'
        info: files promise '/var/cfengine/inputs' repaired
        info: Installing cfe_internal_non_existing_package...
        info: Q:pkg install -y CF_NO ...:Updating FreeBSD repository catalogue...
        info: Q:pkg install -y CF_NO ...:FreeBSD repository is up to date.
        info: Q:pkg install -y CF_NO ...:All repositories are up to date.
        info: Q:pkg install -y CF_NO ...:pkg: No packages available to install matching 'CF_NOMATCH-CF_NOMATCH' have been found in the repositories
        info: Executing 'no timeout' ... '/usr/sbin/sysrc nginx_enable=YES'
      notice: Q: "...sbin/sysrc ngin": nginx_enable: NO -> YES
        info: Last 1 quoted lines were generated by promiser '/usr/sbin/sysrc nginx_enable=YES'
        info: Completed execution of '/usr/sbin/sysrc nginx_enable=YES'
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
        
**It looks like on the first run the services promise only enables the service for starting. Perhaps this could be improved by tweaking the ordering so that service start is delayed until after service enablement has happened so that we don't require a second agent run.**
       
Checking to see if nginx is running after the first execution of cf-agent ...

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # service nginx onestatus
    nginx is not running.
    # Checked to see if nginx is running after the second execution of cf-agent, it's not, sucky
    
Running cf-agent again, will the service state for nginx converge?

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cf-agent -KIf update.cf ; cf-agent -KI
        info: Installing cfe_internal_non_existing_package...
        info: Q:pkg install -y CF_NO ...:Updating FreeBSD repository catalogue...
        info: Q:pkg install -y CF_NO ...:FreeBSD repository is up to date.
        info: Q:pkg install -y CF_NO ...:All repositories are up to date.
        info: Q:pkg install -y CF_NO ...:pkg: No packages available to install matching 'CF_NOMATCH-CF_NOMATCH' have been found in the repositories
        info: Executing 'no timeout' ... '/usr/sbin/service nginx start'
        info: Command related to promiser '/usr/sbin/service nginx start' returned code defined as promise kept 0
      notice: Q: "...in/service ngin": Performing sanity check on nginx configuration:
    Q: "...in/service ngin": nginx: the configuration file /usr/local/etc/nginx/nginx.conf syntax is ok
    Q: "...in/service ngin": nginx: configuration file /usr/local/etc/nginx/nginx.conf test is successful
    Q: "...in/service ngin": Starting nginx.
        info: Last 4 quoted lines were generated by promiser '/usr/sbin/service nginx start'
        info: Completed execution of '/usr/sbin/service nginx start'
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
    
Checking to see if nginx is running after the second execution of cf-agent ...

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # service nginx onestatus
    nginx is running as pid 4131.
    # Checked to see if nginx is running after the second execution of cf-agent, it is, good.
    
Checking to see if service start is executed each time ....

    root@vagrant-host:/usr/local/share/examples/cfengine-masterfiles/masterfiles # cf-agent -KIf update.cf ; cf-agent -KI
        info: Installing cfe_internal_non_existing_package...
        info: Q:pkg install -y CF_NO ...:Updating FreeBSD repository catalogue...
        info: Q:pkg install -y CF_NO ...:FreeBSD repository is up to date.
        info: Q:pkg install -y CF_NO ...:All repositories are up to date.
        info: Q:pkg install -y CF_NO ...:pkg: No packages available to install matching 'CF_NOMATCH-CF_NOMATCH' have been found in the repositories
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
        info: Can not acquire lock for 'nginx' package promise. Skipping promise evaluation
    # Checked to see if service start is executed each time .... it's not, good